### PR TITLE
Tarball ingestion split

### PR DESCRIPTION
**Pull request by siscia**
Different branch to keep the "main branch" untouched.

Here we include all the work done with libarchive, including tests and mocks (we could split them too if necessary).

The next PR will use libarchive as external dependency.

### CODE REVIEW

File: cvmfs/sync_union_tarball.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  /*
+   * Delete the working directories, where the tar is being uncompressed.
+   */
+  // ~SyncUnionTarball();
+
+  /*
+   * Check that the tarball is actually valid and that can be open.
+   * Similarly it creates the base directory or check that it has the ability to
+   * write in it.
+   * Finally it untar the tarball, recursively.
```
Comment (jblomer) [inline cvmfs/sync_union_tarball.h]: This last step should move into the `Traverse()` method, I think.  Let's discuss.

File: cvmfs/sync_union_tarball.cc [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+                                   const std::string &scratch_path,
+                                   const std::string &tarball_path,
+                                   const std::string &base_directory)
+    : SyncUnion(mediator, rdonly_path, union_path, scratch_path),
+      tarball_path_(tarball_path),
+      base_directory_(base_directory) {
+  working_dir_ = "";
+}
+
+/*
```
Comment (jblomer) [inline cvmfs/sync_union_tarball.cc]: This should probably be uncommented or removed.

File: cvmfs/sync_union_tarball.h [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  std::string UnwindWhiteoutFilename(const SyncItem &entry) const;
+  bool IsOpaqueDirectory(const SyncItem &directory) const;
+  bool IsWhiteoutEntry(const SyncItem &entry) const;
+
+ private:
+  const std::string tarball_path_;
+  const std::string base_directory_;
+  std::string working_dir_;
+
+  /*
```
Comment (jblomer) [inline cvmfs/sync_union_tarball.h]: --> `/**` (doxygen syntax)

File: cvmfs/sync_union_tarball.h [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+   * Actually untar the several elements in the tar inside the base directory,
+   * it returns all the recursive tars find in this operation
+   */
+  bool untarPath(const std::string &base_untar_directory_path,
+                 const std::string &tarball_path);
+
+  /*
```
Comment (jblomer) [inline cvmfs/sync_union_tarball.h]: --> `/**`

File: test/unittests/t_sync_union_tarball.cc [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+ */
+
+#include <sync_union_tarball.h>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <util/string.h>
+
+#include <fstream>
```
Comment (jblomer) [inline test/unittests/t_sync_union_tarball.cc]: Generally streams are avoided in cvmfs

File: test/unittests/t_sync_union_tarball.cc [hunk 6]
```diff
@@ -0,0 +1,174 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#include <sync_union_tarball.h>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <util/string.h>
```
Comment (jblomer) [inline test/unittests/t_sync_union_tarball.cc]: --> "util/string.h"

File: test/unittests/t_sync_union_tarball.cc [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+using ::testing::_;
+using ::testing::DefaultValue;
+using ::testing::Return;
+using ::testing::Property;
+
+using ::publish::SyncItem;
+
+namespace {
+
+class T_Sync_union_tarball : public ::testing::Test {
```
Comment (jblomer) [inline test/unittests/t_sync_union_tarball.cc]: --> `T_SyncUnionTarball`

File: test/unittests/t_sync_union_tarball.cc [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+ protected:
+  void SetUp() {
+    zlib::Algorithms algo;
+    DefaultValue<zlib::Algorithms>::Set(algo);
+
+    m_sync_mediator = new publish::MockSyncMediator();
+  }
+
+  std::string CreateTarFile(std::string tar_filename, std::string base64_data) {
```
Comment (jblomer) [inline test/unittests/t_sync_union_tarball.cc]: Add `#include <string>`

Comment (jblomer) [inline test/unittests/t_sync_union_tarball.cc]: I think the parameters can be const references

File: test/unittests/t_sync_union_tarball.cc [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    std::string data_binary;
+    Debase64(base64_data, &data_binary);
+
+    tmp_tar_filename = "/tmp/" + tar_filename;
+
+    tar.open(tmp_tar_filename.c_str(), std::ios::binary);
+
+    if (!tar.is_open()) {
+      cout << "Error: " << strerror(errno);
+      assert(tar.is_open());
```
Comment (jblomer) [inline test/unittests/t_sync_union_tarball.cc]: `#include <cassert>`

File: test/unittests/t_sync_union_tarball.cc [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    assert(tar.write(data_binary.c_str(), data_binary.size()));
```
Comment (jblomer) [inline test/unittests/t_sync_union_tarball.cc]: You should find the necessary convenience routines in util/posix that allow you to not use streams.

File: test/unittests/t_sync_union_tarball.cc [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+    return tmp_tar_filename;
+  }
+
+  virtual void TearDown() {
+    remove(tmp_tar_filename.c_str());
+    delete m_sync_mediator;
+    DefaultValue<zlib::Algorithms>::Clear();
+  }
+
+  publish::MockSyncMediator* m_sync_mediator;
```
Comment (jblomer) [inline test/unittests/t_sync_union_tarball.cc]: --> `sync_mediator_` (like a private member)

File: test/unittests/t_sync_union_tarball.cc [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+  std::string tmp_tar_filename;
```
Comment (jblomer) [inline test/unittests/t_sync_union_tarball.cc]: --> `tmp_tar_filename_`

File: test/unittests/tar_files.h [hunk 13]
```diff
@@ -0,0 +1,268 @@
+/**
```
Comment (jblomer) [inline test/unittests/tar_files.h]: It might make sense to move this file to a sub directory of unittest, like unittest/aux

**Comment by siscia** [comment]
Read it all!

I will work on it tomorrow.

File: cvmfs/sync_union_tarball.cc [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+      case ARCHIVE_OK: {
+        // save the path of the file to extract
+        actuall_path.assign(archive_entry_pathname(entry));
+        // set the new path as base untar directory path + the path of the file
+        archive_entry_set_pathname(
+            entry, (base_untar_directory_path + "/" + actuall_path).c_str());
+
+        result = archive_write_header(dst, entry);
+
+        if (result == ARCHIVE_FATAL) {
```
[the remark this answers is not in the document]
Comment (siscia) [reply cvmfs/sync_union_tarball.cc]: I tried, but I didn't quite like the result, it seemed quite messy.
I opted for a "if cascade" also for distinguish it from the "outer" switch case...

I will come back to it at the end.

File: cvmfs/sync_union_tarball.cc [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+  int result;
+  const void *buff;
+  size_t size;
+  off_t offset;
+
+  while (true) {
+    result = archive_read_data_block(src, &buff, &size, &offset);
+    if (result == ARCHIVE_EOF) return ARCHIVE_OK;
+
+    if (result == ARCHIVE_RETRY || result == ARCHIVE_FATAL) {
```
[the remark this answers is not in the document]
Comment (siscia) [reply cvmfs/sync_union_tarball.cc]: No actually ARCHIVE_RETRY is just meant to indicate that we should retry the operation.
However:
1. Managing the re-try means to wrap everything in a loop and implement basic back pressure (like re-trying 10/20 times and finally abort) and I wasn't quite keen to add this kind of complexity in the code.
2. I really have no idea how this case could happens in moder hardware, I thought that it was some legacy from tape disk that we can most likely safely ignore.

Given the two points above I would suggest to manage it in the simplest way possible and when (and if) this become a problem, we will manage it.

[the remark this answers is not in the document]
Comment (jblomer) [reply cvmfs/sync_union_tarball.cc]: Depending on how libarchive operates internally, "retry" can be a result of an interrupt or a signal which can make normal file I/O return a retry error code in POSIX.  I don't think we need to care about endless loops, but we should wrap archive_read_data_block into our own function that takes care of the errors (probably with a loop to handle `ARCHIVE_RETRY`)

**Comment by siscia** [comment]
@jblomer did I miss any of your comments?

File: CMakeLists.txt [hunk 16]
```diff
@@ -293,6 +293,7 @@ else (NOT MACOSX)
   set (RT_LIBRARY "")
 endif (NOT MACOSX)
 
+
```
Comment (jblomer) [inline CMakeLists.txt]: Just for tidiness, this file could be reverted to its previous state (only blank lines, no need to put in the PR).

File: cvmfs/CMakeLists.txt [hunk 17]
```diff
@@ -246,6 +246,7 @@ set (CVMFS_SWISSKNIFE_SOURCES
   sync_union.cc
   sync_union_overlayfs.cc
   sync_union_aufs.cc
```
Comment (jblomer) [inline cvmfs/CMakeLists.txt]: I only notice now, please re-order alphabetically:

    sync_union_aufs.cc
    sync_union_overlayfs.cc

File: cvmfs/sync_union_tarball.cc [hunk 18]
```diff
@@ -4,30 +4,295 @@
...
 /*
  * Still have to underastand how the madiator, scratch, read and union path
  * works together.
  */
-SyncUnionTarball::SlyncUnionTarball(SyncMediator *mediator,
-                                    const std::string &tarball_path,
-                                    const std::string &base_directory)
-    : tarball_path_(tarball_path), base_directory_(base_directory) {}
+
+// SyncUnionTarball::SyncUnionTarball(VirtualSyncMed)
```
Comment (jblomer) [inline cvmfs/sync_union_tarball.cc]: Can be removed.

File: cvmfs/sync_union_tarball.cc [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+  // flags |= ARCHIVE_EXTRACT_MAC_METADATA;
+  // flags |= ARCHIVE_EXTRACT_SECURE_NOABSOLUTEPATHS;
+
+  src = archive_read_new();
+  assert(src);
+
+  dst = archive_write_disk_new();
+  assert(dst);
+
+  // here we coule also accept _all instead of _tar, however it will require
```
[the remark this answers is not in the document]
Comment (jblomer) [reply cvmfs/sync_union_tarball.cc]: Please note the typo in "coule"

File: cvmfs/sync_union_tarball.h [hunk 20]
```diff
@@ -14,11 +14,21 @@
 #include <set>
 #include <string>
 
+#include "sync_mediator.h"
+
+#include "archive.h"
```
Comment (jblomer) [inline cvmfs/sync_union_tarball.h]: Should go before `sync_mediator.h` (alphabetical ordering)

File: test/unittests/CMakeLists.txt [hunk 21]
```diff
@@ -102,6 +102,7 @@ set(CVMFS_UNITTEST_FILES
   t_statistics.cc
   t_swissknife_lease.cc
   t_synchronizing_counter.cc
+  t_sync_union_tarball.cc
```
Comment (jblomer) [inline test/unittests/CMakeLists.txt]: Better before `t_synchronizing_counter`

File: test/unittests/CMakeLists.txt [hunk 22]
```diff
@@ -220,6 +221,10 @@ set (CVMFS_UNITTEST_SOURCES
   ${CVMFS_SOURCE_DIR}/swissknife_history.cc
   ${CVMFS_SOURCE_DIR}/swissknife_lease_json.cc
   ${CVMFS_SOURCE_DIR}/swissknife_lease_curl.cc
+  ${CVMFS_SOURCE_DIR}/sync_union.cc
```
Comment (jblomer) [inline test/unittests/CMakeLists.txt]: Please re-order.

File: test/unittests/t_sync_union_tarball.cc [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+ * This file is part of the CernVM File System.
+ */
+
+#include <sync_union_tarball.h>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <cassert>
+#include <iostream>
```
Comment (jblomer) [inline test/unittests/t_sync_union_tarball.cc]: Please remove.

File: test/unittests/t_sync_union_tarball.cc [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+#include <string>
+
+#include "mock/m_sync_mediator.h"
```
Comment (jblomer) [inline test/unittests/t_sync_union_tarball.cc]: Combine with following includes and re-order alphabetically.

**Comment by siscia** [comment]
@jblomer @radupopescu Can we please retry the mac build? It seems a connection error while trying to download the dependencies.

**Pull request closed without merging**